File: questions/实现一个正常的栈增加一个返回min的方法/implement.js
/**
 * 栈的实现
 */
class Stack {
    constructor() {
        this.item = []
    }
    push(value) {
        this.item.push(value)
    }
    pop() {
        return this.item.pop()
    }
    peek() {
        return this.item[this.item.length-1]
    }
    isEmpty() {
        return this.item.length === 0
    }
}

/**
 * 目标栈
 * 多了一个getMin方法
 */
class TargetStack {
    constructor() {
        this.stack = new Stack()
        this.minStack = new Stack()
        this.min = null
    }

    push(value) {
        this.stack.push(value)
        if (!this.min || this.min > value) {
            this.min = value
        }
        this.minStack.push(this.min)
    }

    pop() {
        this.minStack.pop()
        return this.stack.pop()
    }

    getMin() {
        this.stack.pop()
        return this.minStack.pop()
    }
}

File: questions/实现一个正常的栈增加一个返回min的方法/implement.java
package 实现一个正常的栈增加一个返回min的方法;

import java.util.Stack;

/**
 * 为一个正常的栈实现一个getMin方法
 * <p>时间 2020/8/14 - 1:53 上午</p>
 *
 * @author 吕蓬
 * @version 1.0
 */
public class implement {
    private Stack <Integer> stackData;
    private Stack <Integer> stackMin;

    public static void main(String[] args) {

    }
    public void push(int value) {
        if(this.stackMin.isEmpty()) {
            this.stackMin.push(value);
        } else if (value < this.stackMin.peek()) {
            this.stackMin.push(value);
        } else {
            this.stackMin.push(this.stackMin.peek());
        }
        this.stackData.push(value);
    }
    public int pop() {
        if (this.stackData.isEmpty()) {
            throw new Error("Underflow");
        }
        this.stackMin.pop();
        return this.stackData.pop();
    }
    public int getMin() {
        if (this.stackData.isEmpty()) {
            throw new Error("Underflow");
        }
        this.stackData.pop();
        return this.stackMin.pop();
    }
    public boolean isEmpty() {
        return this.stackData.isEmpty();
    }
}
